core/state/snapshot/generate.go

```diff
@@ -159,169 +178,289 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
 	rawdb.WriteSnapshotGenerator(db, blob)
 }
 
+// proveRange proves the state segment with particular prefix is "valid".
+// The iteration start point will be assigned if the iterator is restored from
+// the last interruption. Max will be assigned in order to limit the maximum
+// amount of data involved in each iteration.
+func (dl *diskLayer) proveRange(root common.Hash, tr *trie.SecureTrie, prefix []byte, kind string, origin []byte, max int, onValue func([]byte) ([]byte, error)) ([][]byte, [][]byte, []byte, bool, error) {
+	var (
+		keys  [][]byte
+		vals  [][]byte
+		count int
+		last  []byte
+		proof = rawdb.NewMemoryDatabase()
+		iter  = dl.diskdb.NewIterator(prefix, origin)
```

Comment:
I don't think you ever release this iterator...? It needs to be released before any return, and you probably should do it right after the for-loop rather than deferring it

Reply:
This is a bit problematic. You will have one outer iterator, whih does some work, and calls `OnAccount`, which starts another iterator for storage slots, and will operate on that until it's done, and only then release the outer iterator and allow it to move on. 
Not sure about the semantics here, and what needs to happen when. Would it be possible to

-  first iterate through the keys, soup up the values, close the iterator. 
- then iterate through the key/values, invoke the callbacks

Then each iterator would only be 'live' while we read out `max` items, instead of these long-lived db iterators. 

?

Reply:
No, it's a bug :) We should release the db handler. All the entries will be returned to outside, so it's totally fine(and should) to release the iterator here.